### PR TITLE
Fix S4035 FP on abstract declarations

### LINKED ISSUE #2322: Rule  S4035: False positive for abstract declarations
### Description

```csharp
public abstract class ValueObject<T> : IEquatable<T> where T : ValueObject<T>
{
    public abstract bool Equals(T other); // compliant
}

public class MyValueObject : ValueObject<MyValueObject>
{
    public override bool Equals(MyValueObject other)
    {
        return RefrenceEquals(this, other); //non-compliant
    }
}

public sealed class MySealedValueObject : ValueObject<MySealedValueObject >
{
    public override bool Equals(MySealedValueObject  other)
    {
        return RefrenceEquals(this, other); //compliant
    }
}
```

To me, this should not raise the this concern on the abstract class with an abstract method. The implementing class should, if not sealed.

### Related information
* #461
* 7.11.0.8083
* Visual Studio 15.9.7

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**
Fixes #2322

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EquatableClassShouldBeSealed.cs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-/*
+﻿/*
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EquatableClassShouldBeSealed.cs]: nitpick - this is some weird artefact, you should be able to revert this line

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/EquatableClassShouldBeSealed.cs [hunk 2]
```diff
@@ -7,6 +7,11 @@ public class ClassWithVirtualEquals : IEquatable<ClassWithVirtualEquals> // Comp
         public virtual bool Equals(ClassWithVirtualEquals other) => true;
     }
 
+    public abstract class ClassWithAbstractEquals : IEquatable<ClassWithAbstractEquals> // Compliant
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/EquatableClassShouldBeSealed.cs]: what about an `abstract` class with a non-abstract `Equals` method?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Utilities/Rules.Description/S4035.html [hunk 3]
```diff
@@ -4,7 +4,7 @@
 <p>Alternatively <code>IEqualityComparer&lt;T&gt;</code> provides a safer interface and is used by collections or <code>Equals</code> could be made
 <code>virtual</code>.</p>
 <p>This rule raises an issue when a unsealed, <code>public</code> or <code>protected</code> class implements <code>IEquitable&lt;T&gt;</code> and the
-<code>Equals</code> is not <code>virtual</code>.</p>
+<code>Equals</code> is neither <code>virtual</code> nor <code>abstract</code>.</p>
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Utilities/Rules.Description/S4035.html]: running `rspec cs S4035` should also modify `sonaranalyzer-dotnet/rspec/cs/S4035_c#.html`

Comment (andrea-guarino-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.Utilities/Rules.Description/S4035.html]: Indeed, for some reason `rspec` script is modifying more files, so I had to select them manually and I missed `S4035_c#.html`. Thanks!

Comment (andrei-epure-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.Utilities/Rules.Description/S4035.html]: I opened #2572 to track this

**Comment by Corniel** [comment]
LGTM. :)

**Pull request merged**